src/libraries/System.Text.Json/tests/JsonNode/DynamicTests.cs

```diff
@@ -0,0 +1,286 @@
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json.Serialization;
+using Microsoft.CSharp.RuntimeBinder;
+using Xunit;
+
+namespace System.Text.Json.Node.Tests
+{
+    public static class DynamicTests
+    {
+        [Fact]
+        public static void ImplicitOperators()
+        {
+            dynamic jObj = new JsonObject();
+
+            // Dynamic objects do not support object initializers.
+
+            // Primitives
+            jObj.MyString = "Hello!";
+            Assert.IsAssignableFrom<JsonValue>(jObj.MyString);
+
+            jObj.MyNull = null;
+            jObj.MyBoolean = false;
+
+            // Nested array
+            jObj.MyArray = new JsonArray(2, 3, 42);
+
+            // Additional primitives
+            jObj.MyInt = 43;
+            jObj.MyDateTime = new DateTime(2020, 7, 8);
+            jObj.MyGuid = new Guid("ed957609-cdfe-412f-88c1-02daca1b4f51");
+
+            // Nested objects
+            jObj.MyObject = new JsonObject();
+            jObj.MyObject.MyString = "Hello!!";
+
+            jObj.Child = new JsonObject();
+            jObj.Child.ChildProp = 1;
+
+            var options = new JsonSerializerOptions();
+            options.UnknownTypeHandling = JsonUnknownTypeHandling.JsonNode;
+
+            string json = jObj.ToJsonString(options);
+            JsonTestHelper.AssertJsonEqual(JsonNodeTests.ExpectedDomJson, json);
+        }
+
+        private enum MyCustomEnum
+        {
+            Default = 0,
+            FortyTwo = 42,
+            Hello = 77
+        }
+
+        [Fact]
+        public static void Primitives_UnknownTypeHandling()
+        {
+            var options = new JsonSerializerOptions();
+            options.UnknownTypeHandling = JsonUnknownTypeHandling.JsonNode;
+            options.Converters.Add(new JsonStringEnumConverter());
+
+            dynamic obj = JsonSerializer.Deserialize<object>(Serialization.Tests.DynamicTests.Json, options);
+            Assert.IsAssignableFrom<JsonObject>(obj);
+
+            // JsonValue created from a JSON string.
+            Assert.IsAssignableFrom<JsonValue>(obj.MyString);
+            Assert.Equal("Hello", (string)obj.MyString);
+
+            // Verify other string-based types.
+            // Since this requires a custom converter, an exception is thrown.
+            Assert.ThrowsAny<RuntimeBinderException>(() => (MyCustomEnum)obj.MyString);
+
+            Assert.Equal(Serialization.Tests.DynamicTests.MyDateTime, (DateTime)obj.MyDateTime);
+            Assert.Equal(Serialization.Tests.DynamicTests.MyGuid, (Guid)obj.MyGuid);
+
+            // JsonValue created from a JSON bool.
+            Assert.IsAssignableFrom<JsonValue>(obj.MyBoolean);
+            bool b = (bool)obj.MyBoolean;
+            Assert.True(b);
+
+            // Numbers must specify the type through a cast or assignment.
+            Assert.IsAssignableFrom<JsonValue>(obj.MyInt);
+            Assert.ThrowsAny<Exception>(() => obj.MyInt == 42L);
```

Comment:
Changed to `RuntimeBinderException`. We could catch and throw InvalidOperation in this case; I'll look at that for consistency sake with non-dynamic.

Yes you can change the value with `obj.MyInt = "Hello"` since that uses the indexer to create a new node, and then replace the previous node.

Reply:
+1 to consistency with non-dynamic.

Reply:
I'll address this after the PR. To support this will require changes to the lower-level dynamic interop in `MetaDynamic.cs` and adding more extensive testing for failure cases.